src/qforte/utils/qubit_tapering.py

```diff
@@ -0,0 +1,345 @@
+    # 1) Construct the parity check matrix E from a given qubit Hamiltonian
+    # 2) Find the basis of the kernel of the parity check matrix, ker{E}
+    # 3) Find maximal Abelian subgroup of the basis of ker{E}
+    # 4) Construct unitary matrices that will be used in the qubit_tapering algorithm
+    # 5) Compute the various signs corresponding to the eigenvalues of the sigma_x operators that will be removed
+
+    # 1) Construct the parity check matrix E from a given qubit Hamiltonian
+
+    n_qubits = hamiltonian.num_qubits()
+    n_strings = len(hamiltonian.terms())
+
+    ## The rows of the parity check matrix correspond to Pauli strings of the qubit Hamiltonian.
+    ## The columns of the parity check matrix correspond to the Z_(n_qubits - 1), ..., Z_0, X_(n_qubits - 1), ..., X_0 Pauli gates.
+    prt_chck_mtrx = np.zeros((n_strings, 2 * n_qubits), dtype=np.uint8)
+    for i, (_, string) in enumerate(hamiltonian.terms()):
+        for pauli in string.gates():
+            XYZ_gate = pauli.gate_id()
+            trgt = pauli.target()
+            if XYZ_gate in ['Y', 'Z']:
+                prt_chck_mtrx[i, n_qubits - trgt - 1] = 1
+            if XYZ_gate in ['X', 'Y']:
+                prt_chck_mtrx[i, 2 * n_qubits - trgt - 1] = 1
+
+    # 2) Find a basis of the kernel of the parity check matrix
+    # See https://en.wikipedia.org/wiki/Kernel_(linear_algebra)#Computation_by_Gaussian_elimination
+
+    ## Augment the (n_strings) × (2 * n_qubits) parity check matrix by the (2 * n_qubits) × (2 * n_qubits) identity matrix
+
+    prt_chck_mtrx_aug = np.concatenate((prt_chck_mtrx, np.identity(2 * n_qubits, dtype=np.uint8)), axis=0)
+
+    ## Find column echelon form (CEF) of augmented matix.
+
+    ### We only work with the columns of the parity check matrix that have not been transformed yet to CEF.
+    ### This is controlled by the "idx" variable
+    idx = 0
+
+    for row in range(n_strings):
+        ### Find a non-zero row
+        nonzero_indices = np.argwhere(prt_chck_mtrx_aug[row,idx:])
+        if nonzero_indices.shape[0] > 0:
+            column_to_be_CEFed = int(nonzero_indices[0]) + idx
+            ### Bring first non-zero element in this row immediately to the right of that of the previous row
+            prt_chck_mtrx_aug[:, [column_to_be_CEFed , idx]] = prt_chck_mtrx_aug[:, [idx, column_to_be_CEFed]]
+            ### Eliminate remaining non-zero row elements in the part of the parity check matrix that is not in CEF
+            for i in range(1,nonzero_indices.shape[0]):
+                column_not_in_CEF = int(nonzero_indices[i]) + idx
+                prt_chck_mtrx_aug[:,column_not_in_CEF] = np.bitwise_xor(prt_chck_mtrx_aug[:, idx], prt_chck_mtrx_aug[:,column_not_in_CEF])
+            ### decrease the number of columns that need to be checked
+            idx += 1
+
+    ## Find basis vectors of kernel by checking for zero columns in the CEF of the parity check matrix
+
+    dim = False
```

Comment:
`dim` is not descriptive and you don't use the value of `basis` you initialize below. Can you initialize `basis` with shape `(0, 2 * n_qubits)`? Then you can always concatenate `new_basis_vector` to `basis` and don't need `dim`.